File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int KthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stack = new();
        TreeNode curr = root;
        int count = 0;

        while (stack.Count > 0 || curr is not null) {
            while (curr is not null) {
                stack.Push(curr);
                curr = curr.left;
            }

            curr = stack.Pop();
            if (count == k-1) {
                return curr.val;
            }
            count++;

            curr = curr.right;
        }
        return curr.val;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        # in order BST always produces increasing list
        prev = float('-inf')
        stack = []
        curr = root

        while stack or curr:
            # left
            while curr:
                stack.append(curr)
                curr = curr.left

            #parent
            curr = stack.pop()
            if curr.val <= prev:
                return False
            
            #right
            prev = curr.val
            curr = curr.right
        
        return True


        
    